KAFKA-19160;KAFKA-19164; Improve performance of fetching stable offsets

squah-confluent opened the pull request:
When fetching stable offsets in the group coordinator, we iterate over
all requested partitions. For each partition, we iterate over the
group's ongoing transactions to check if there is a pending
transactional offset commit for that partition.

This can get slow when there are a large number of partitions and a
large number of pending transactions. Instead, maintain a list of
pending transactions per partition to speed up lookups.

Reviewers: Shaan, Dongnuo Lyu <dlyu@confluent.io>, Chia-Ping Tsai
 <chia7712@gmail.com>, David Jacot <djacot@confluent.io>

squah-confluent committed:
KAFKA-19160: Improve performance of fetching stable offsets

When fetching stable offsets in the group coordinator, we iterate over
all requested partitions. For each partition, we iterate over the
group's ongoing transactions to check if there is a pending
transactional offset commit for that partition.

This can get slow when there are a large number of partitions and a
large number of pending transactions. Instead, maintain a list of
pending transactions per partition to speed up lookups.

squah-confluent commented on group-coordinator/src/main/java/org/apache/kafka/coordinator/group/OffsetMetadataManager.java:
```diff
@@ -194,9 +194,16 @@ public OffsetMetadataManager build() {
      * The open transactions (producer ids) keyed by group.
+     * Tracks whether groups have any open transactions.
      */
     private final TimelineHashMap<String, TimelineHashSet<Long>> openTransactionsByGroup;
```
I wanted to replace this map entirely, but we use it in `cleanupExpiredOffsets` to check whether a group has any pending transactions.

dajac replied:
Can we use openTransactionsByGroupTopicAndPartition for this too? If the group is present in openTransactionsByGroupTopicAndPartition, it means that it has an open transaction.

squah-confluent replied:
We clear entries from that map when offsets are deleted non-transactionally. If all offsets in a transaction are deleted, then there can be an open transaction for a group without a record of it in openTransactionsByGroupTopicAndPartition.

squah-confluent replied:
I've removed the 2nd map of group ids -> producer ids. Now we only track (group id, topic, partition) -> producer ids. As a side effect, it fixes KAFKA-19164.

shaan150 commented on group-coordinator/src/main/java/org/apache/kafka/coordinator/group/OffsetMetadataManager.java:
```diff
@@ -650,24 +658,18 @@ public int deleteAllOffsets(
-                }
+        TimelineHashMap<String, TimelineHashMap<Integer, TimelineHashSet<Long>>> openTransactionsByTopic =
+            openTransactionsByGroupTopicAndPartition.get(groupId);
+        if (openTransactionsByTopic != null) {
```
There are three nested for loops, this can be avoided.

Something like the following might help (was done for quickness)

```java
var openTransactionsByTopic = openTransactionsByGroupTopicAndPartition.get(groupId);
if (openTransactionsByTopic == null) return;

for (var topicEntry : openTransactionsByTopic.entrySet()) {
    String topic = topicEntry.getKey();
    var openTransactionsByPartition = topicEntry.getValue();

    for (var partitionEntry : openTransactionsByPartition.entrySet()) {
        int partition = partitionEntry.getKey();

        // Single check per partition
        if (!hasCommittedOffset(groupId, topic, partition)) {
            records.add(GroupCoordinatorRecordHelpers.newOffsetCommitTombstoneRecord(groupId, topic, partition));
            numDeletedOffsets.getAndIncrement();
        }
    }
}
```

squah-confluent replied:
I missed this when refactoring the code, thank you. The previous code also has this bug, except it was not as obvious. I've fixed it now.

shaan150 commented on group-coordinator/src/main/java/org/apache/kafka/coordinator/group/OffsetMetadataManager.java:
```diff
@@ -1005,21 +1005,41 @@ public void replay(
                     }
-                });
+                }
             }
```
This whole area is pretty badly structured, and maintainability will prove difficult, potentially fragile. 

I would split this out into methods, and tidy some of it up. This is an example on how i'd put it, this code isn't tested baring mind and is more of a boiler plate example:

```java
    public void clearOpenTransactions(final String groupId, final String topic, inal int partition) {
        final TimelineHashMap<Integer, TimelineHashSet<Long>> partitionMap = getPartitionMap(groupId, topic);
        if (partitionMap == null) return;
    
        final TimelineHashSet<Long> openProducerIds = partitionMap.get(partition);
        if (openProducerIds == null) return;
    
        removePendingOffsets(openProducerIds, groupId, topic, partition);
    
        partitionMap.remove(partition);
        cleanupIfEmpty(partitionMap, getTopicMap(groupId), topic);
        cleanupIfEmpty(getTopicMap(groupId), openTransactionsByGroupTopicAndPartition, groupId);
    }
    
    private TimelineHashMap<Integer, TimelineHashSet<Long>> getPartitionMap(final String groupId, final String topic) {
        final TimelineHashMap<String, TimelineHashMap<Integer, TimelineHashSet<Long>>> topicMap =
            openTransactionsByGroupTopicAndPartition.get(groupId);
        if (topicMap == null) return null;
        return topicMap.get(topic);
    }
    
    private TimelineHashMap<String, TimelineHashMap<Integer, TimelineHashSet<Long>>> getTopicMap(final String groupId) {
        return openTransactionsByGroupTopicAndPartition.get(groupId);
    }
    
    private void removePendingOffsets(
        final Set<Long> producerIds, 
        final String groupId, 
        final String topic, 
        final int partition) {
        for (final Long producerId : producerIds) {
            final Offsets offsets = pendingTransactionalOffsets.get(producerId);
            if (offsets != null) {
                offsets.remove(groupId, topic, partition);
            }
        }
    }
    
    private <K, V extends Map<?, ?>> void cleanupIfEmpty(final V innerMap, final Map<K, V> outerMap, final K key) {
        if (innerMap != null && innerMap.isEmpty()) {
            outerMap.remove(key);
        }
    }
```

shaan150 commented on group-coordinator/src/main/java/org/apache/kafka/coordinator/group/OffsetMetadataManager.java:
```diff
@@ -1043,14 +1064,39 @@ public void replayEndTransactionMarker(
+                    }
+                });
+            });
         });
```
Previous point stands, how i'd do it would be something like the below attempt, the cleanupIfEmpty method from earlier here could be used if implemented 

```java

public void clearOpenTransactionsForProducer(final long producerId, final PendingOffsets pendingOffsets) {
    for (final Map.Entry<String, Map<String, Map<Integer, OffsetAndMetadata>>> groupEntry : pendingOffsets.offsetsByGroup.entrySet()) {
        final String groupId = groupEntry.getKey();
        final Map<String, Map<Integer, OffsetAndMetadata>> topicOffsets = groupEntry.getValue();

        removeProducerFromGroupTransactions(groupId, producerId);

        final TimelineHashMap<String, TimelineHashMap<Integer, TimelineHashSet<Long>>> topicMap =
            openTransactionsByGroupTopicAndPartition.get(groupId);
        if (topicMap == null) continue;

        processTopicOffsets(producerId, groupId, topicOffsets, topicMap);
        cleanupIfEmpty(topicMap, openTransactionsByGroupTopicAndPartition, groupId);
    }
}

private void removeProducerFromGroupTransactions(final String groupId, final long producerId) {
    final TimelineHashSet<Long> groupTransactions = openTransactionsByGroup.get(groupId);
    if (groupTransactions == null) return;

    groupTransactions.remove(producerId);
    if (groupTransactions.isEmpty()) {
        openTransactionsByGroup.remove(groupId);
    }
}

private void processTopicOffsets(
    final long producerId,
    final String groupId,
    final Map<String, Map<Integer, OffsetAndMetadata>> topicOffsets,
    final TimelineHashMap<String, TimelineHashMap<Integer, TimelineHashSet<Long>>> topicMap) {

    for (final Map.Entry<String, Map<Integer, OffsetAndMetadata>> topicEntry : topicOffsets.entrySet()) {
        final String topic = topicEntry.getKey();
        final Map<Integer, OffsetAndMetadata> partitionOffsets = topicEntry.getValue();

        final TimelineHashMap<Integer, TimelineHashSet<Long>> partitionMap = topicMap.get(topic);
        if (partitionMap == null) continue;

        for (final Integer partitionId : partitionOffsets.keySet()) {
            removeProducerFromPartitionMap(producerId, partitionId, partitionMap);
        }

        cleanupIfEmpty(partitionMap, topicMap, topic);
    }
}

private void removeProducerFromPartitionMap(
    final long producerId,
    final int partitionId,
    final TimelineHashMap<Integer, TimelineHashSet<Long>> partitionMap) {

    final TimelineHashSet<Long> partitionTransactions = partitionMap.get(partitionId);
    if (partitionTransactions == null) return;

    partitionTransactions.remove(producerId);

    if (partitionTransactions.isEmpty()) {
        partitionMap.remove(partitionId);
    }
}
```

shaan150 commented on group-coordinator/src/main/java/org/apache/kafka/coordinator/group/OffsetMetadataManager.java:
```diff
@@ -194,9 +194,16 @@ public OffsetMetadataManager build() {
+     * The open transactions (producer ids) keyed by group id, topic name and partition id.
+     * Tracks whether partitions have any pending transactional offsets.
+     */
+    private final TimelineHashMap<String, TimelineHashMap<String, TimelineHashMap<Integer, TimelineHashSet<Long>>>> openTransactionsByGroupTopicAndPartition;
```
I understand your reasoning behind this addition, but it does introduce significant complexity and carries a high risk of data duplication if not carefully managed. While I appreciate this may serve as a stop-gap for now, I do think it’s important that this is revisited going forward. A more maintainable long-term solution, perhaps wrapping this logic in a dedicated structure or abstraction, would help reduce coupling and make it easier to evolve the logic cleanly.

shaan150 reviewed (commented):
Left a few comments, it's mostly due to where the file was when picked up. That said, changes in areas like this are a good opportunity to improve structure as well. I get that it might add a bit of time now, but it'll pay off long-term by reducing complexity and easing future changes

dajac commented:
@squah-confluent Thanks for the patch. Could we write a micro benchmark to demonstrate the gain?

chia7712 commented on group-coordinator/src/main/java/org/apache/kafka/coordinator/group/OffsetMetadataManager.java:
```diff
@@ -650,24 +658,18 @@ public int deleteAllOffsets(
+        if (openTransactionsByTopic != null) {
+            openTransactionsByTopic.forEach((topic, openTransactionsByPartition) -> {
+                openTransactionsByPartition.forEach((partition, producerIds) -> {
+                    producerIds.forEach(producerId -> {
```
Excuse me, why to iterate all producerIds if we don't use it actually in creating tombstone?

squah-confluent replied:
I missed this when refactoring the code, thank you. The previous code also has this bug, except it was not as obvious. I've fixed it now.

squah-confluent committed:
KAFKA-19160: Add benchmark for fetching stable offsets

squah-confluent committed:
fixup: remove mapping of groups to open transactions

It was kept around in the initial PR to avoid deletion of groups with
open transactions, but it's okay to delete groups with open transactions
without any offsets, since committing the transaction is a no-op. If the
client tries to add more transactional offsets to a deleted group, we
may either recreate the group or return an error depending on the
generation in the request.

squah-confluent committed:
fixup: don't generate extra tombstones when deleting offsets

squah-confluent committed:
fixup: refactor

squah-confluent commented:
> @squah-confluent Thanks for the patch. Could we write a micro benchmark to demonstrate the gain?

Added a benchmark.
Before:
```
Benchmark                              (partitionCount)  (transactionCount)  Mode  Cnt    Score    Error  Units
TransactionalOffsetFetchBenchmark.run              4000                4000  avgt    5  452.957 ± 7.883  ms/op
```
After:
```
Benchmark                              (partitionCount)  (transactionCount)  Mode  Cnt  Score   Error  Units
TransactionalOffsetFetchBenchmark.run              4000                4000  avgt    5  0.196 ± 0.005  ms/op
```

squah-confluent committed:
fixup: clean up benchmark code

squah-confluent commented:
@shaan150 I decided to do something a little different and factored out the code into `Map` interface-like operations.

squah-confluent committed:
fixup: clean up benchmark code

dongnuo123 commented on group-coordinator/src/main/java/org/apache/kafka/coordinator/group/OffsetMetadataManager.java:
```diff
@@ -192,10 +193,165 @@ public OffsetMetadataManager build() {
+         * @param partition The partition.
+         * @return {@code true} if the given group has any pending transactional offsets for the given topic and partition.
+         */
+        private boolean contains(String groupId, String topic, int partition) {
+            TimelineHashSet<Long> openTransactions = get(groupId, topic, partition);
+            return openTransactions != null;
+        }
```
Could we do something like
```
        TimelineHashMap<String, TimelineHashMap<Integer, TimelineHashSet<Long>>> topicMap = 
            openTransactionsByGroup.get(groupId);
            
        if (topicMap == null) return false;
        
        TimelineHashMap<Integer, TimelineHashSet<Long>> partitionMap = topicMap.get(topic);
        return partitionMap != null && partitionMap.containsKey(partition);
```
to avoid extra lookup?

squah-confluent replied:
I checked and `containsKey` does basically the same operation as a `get`.
https://github.com/apache/kafka/blob/08f6042f7a24753fc96389c6cf4c6042999bbd39/server-common/src/main/java/org/apache/kafka/timeline/TimelineHashMap.java#L120-L126
https://github.com/apache/kafka/blob/08f6042f7a24753fc96389c6cf4c6042999bbd39/server-common/src/main/java/org/apache/kafka/timeline/TimelineHashMap.java#L139-L150

dongnuo123 commented on group-coordinator/src/main/java/org/apache/kafka/coordinator/group/OffsetMetadataManager.java:
```diff
@@ -192,10 +193,165 @@ public OffsetMetadataManager build() {
+         *
+         * Values in each level of the map will never be empty collections.
+         */
+        private final TimelineHashMap<String, TimelineHashMap<String, TimelineHashMap<Integer, TimelineHashSet<Long>>>> openTransactionsByGroup;
```
The current 3-layer map has already improved the performance significantly. If we are still not happy with the perf maybe we can consider flattening the map to (group id, topic, partition) -- producer id set

squah-confluent replied:
I tried a couple of approaches and settled on inlining `hasPendingTransactionalOffsets` into the offset fetch path. This way we don't do string comparisons of the group id and topic name for every partition and also avoid allocations from a compound key.

```
Benchmark                              (partitionCount)  (transactionCount)  Mode  Cnt  Score   Error  Units
TransactionalOffsetFetchBenchmark.run              4000                4000  avgt    5  0.129 ± 0.002  ms/op
```

squah-confluent committed:
fixup: inline hasPendingTransactionalOffset lookups into offset fetch paths

squah-confluent committed:
fixup: add missing javadoc for openTransactions field

dajac commented on group-coordinator/src/main/java/org/apache/kafka/coordinator/group/OffsetMetadataManager.java:
```diff
@@ -736,6 +871,11 @@ public OffsetFetchResponseData.OffsetFetchResponseGroup fetchOffsets(
+        // We inline the lookups from hasPendingTransactionalOffsets here, to avoid repeating string
+        // comparisons of group ids and topic names for every partition. They're only used when the
+        // client has requested stable offsets.
+        final TimelineHashMap<String, TimelineHashMap<Integer, TimelineHashSet<Long>>> openTransactionsByTopic =
+            requireStable ? openTransactions.openTransactionsByGroup.get(request.groupId()) : null;
```
It feels a bit weird that we don't take `lastCommittedOffset` into consideration here. It works because `lastCommittedOffset` is `MAX_VALUE` when `requireStable` is `true` but it still feel weird. Have you considered using `lastCommittedOffset` in order to be consistent with `groupOffsets`?

squah-confluent replied:
I think that's reasonable, I've updated the PR.

dajac commented on group-coordinator/src/main/java/org/apache/kafka/coordinator/group/OffsetMetadataManager.java:
```diff
@@ -802,11 +946,18 @@ public OffsetFetchResponseData.OffsetFetchResponseGroup fetchAllOffsets(
+        // comparisons of group ids and topic names for every partition. They're only used when the
+        // client has requested stable offsets.
+        final TimelineHashMap<String, TimelineHashMap<Integer, TimelineHashSet<Long>>> openTransactionsByTopic =
+            requireStable ? openTransactions.openTransactionsByGroup.get(request.groupId()) : null;
```
ditto.

dajac reviewed (commented):
@squah-confluent Thanks for the patch. I left a few comments for consideration.

squah-confluent committed:
fixup: query open transactions at lastCommittedOffset

dajac reviewed (approved):
lgtm, thanks
